cheroot/server.py

```diff
@@ -340,8 +341,65 @@ def __next__(self):
 
     next = __next__
 
+class RFile:
+    def __init__(self, rfile):
+        self.rfile = rfile
+
+    def read(self, size=None):
+        raise NotImplementedError
+
+    def readline(self, size=None):
+        raise NotImplementedError
+
+    def readlines(self, sizehint=None):
+        raise NotImplementedError
+
+    def close(self):
+        self.rfile.close()
+
+class HyperRFile(RFile):
+
+    def __init__(self, rfile, conn):
+        super(HyperRFile, self).__init__(rfile)
+        self.conn = conn
+
+    def read(self, size=None):
+        if self.conn.their_state != h11.SEND_BODY or size == 0:
+            return b""
+        event = self.conn.next_event()
+        if event is h11.NEED_DATA:
+            self.conn.receive_data(self.rfile.read(size))
+            event = self.conn.next_event()
+        assert isinstance(event, h11.Data)
```

Comment:
Is this some sort of an exit-check for ensuring contracts?

Reply:
Anything other than `h11.Data` here should be undefined behavior. Currently, I believe the only possible way we don't is if we get an EndOfMessage, but I'm pretty sure that were that to happen, h11 would raise an exception. I believe I saw this in the example code that they provide, so I kept it in for fun. 

I can't really think of any other way it would be anything else, so I kept it in as a general catch-all.

Reply:
Alright. This is a practice people follow when doing contract programming just as a sanity-check. `assert`s are stripped off in optimized mode. But it's useful to keep this as an exit-check.